protocols/autonat/src/v2/client/handler/dial_back.rs

```diff
@@ -90,3 +88,50 @@ impl ConnectionHandler for Handler {
         }
     }
 }
+
+struct State {
+    stream: libp2p_swarm::Stream,
+    oneshot: Option<oneshot::Receiver<io::Result<()>>>,
+}
+
+#[derive(Debug)]
+pub struct IncomingNonce {
+    pub nonce: Nonce,
+    pub sender: oneshot::Sender<io::Result<()>>,
+}
+
+fn perform_dial_back(
+    stream: libp2p_swarm::Stream,
+) -> impl futures::Stream<Item = io::Result<IncomingNonce>> {
+    let state = State {
+        stream,
+        oneshot: None,
+    };
+    futures::stream::unfold(state, |mut state| async move {
+        if let Some(ref mut receiver) = state.oneshot {
+            if receiver.await.is_err() {
+                return Some((Err(io::Error::from(io::ErrorKind::Other)), state));
+            }
+            if let Err(e) = protocol::dial_back_response(&mut state.stream).await {
+                let _ = state.stream.close().await;
+                return Some((Err(e), state));
+            }
+            if let Err(e) = state.stream.close().await {
+                return Some((Err(e), state));
+            }
+            return None;
+        }
+
+        let nonce = match protocol::recv_dial_back(&mut state.stream).await {
+            Ok(nonce) => nonce,
+            Err(err) => {
+                let _ = state.stream.close().await;
```

Comment:
I don't think we need to attempt to `close` on error. In fact, I think we should reserve `close` for the happy path.

Reply:
That's a good point. I hope I changed it.

Reply:
yeah that's a good point. I changed that